File: 295-FindMedianfromDataStream/295-FindMedianfromDataStream.cpp
// Inserting into a std::set does not invalidate the iterators since it uses a balanced bst.

class MedianFinder {
public:
    map<int, vector<int>> numMap;
    map<int, vector<int>>::iterator currIter;
    int mapIndex = 0;
    int mapSize = 0; 

    int currIndex = -1;

    MedianFinder() {
        
    }

    // 0100
    
    void addNum(int num) {
        mapSize++;

        numMap[num].push_back(numMap[num].size());
        if (currIndex == -1) {
            currIndex = 0;
            currIter = numMap.begin();
        } 
        if (num < currIter->first) {
            mapIndex++;
        }
    }
    
    double findMedian() {
        cout << "findMedian" << endl;
        cout << mapIndex << " " << (mapSize / 2.0) << " " << mapSize << endl;
        while (mapIndex < (mapSize / 2.0) - 1) {
            mapIndex++;
            currIndex++;
            if (currIndex > currIter->second.size() - 1) {
                currIter++;
                currIndex = 0;
            }
        }
        while (mapIndex >= (mapSize / 2.0)) {
            mapIndex--;
            currIndex--;
            if (currIndex < 0) {
                currIter--;
                currIndex = currIter->second.size() - 1;
            }
        }
        cout << mapIndex << " " << (mapSize / 2.0) << " " << mapSize << endl;
        // cout << currIter->first << " " << currIndex << endl;
        if (mapSize % 2 == 0) {
            auto iterCopy = currIter;
            int indexCopy = currIndex;
            indexCopy++;
            if (indexCopy > iterCopy->second.size() - 1) {
                iterCopy++;
            }
            return (currIter->first + iterCopy->first) / 2.0;
        } else {
            return currIter->first;
        }

        // -1, -2, -3, -4, -5
    }
};

/**
 * Your MedianFinder object will be instantiated and called as such:
 * MedianFinder* obj = new MedianFinder();
 * obj->addNum(num);
 * double param_2 = obj->findMedian();
 */

File: 295-FindMedianfromDataStream/295-FindMedianfromDataStream.py
# Last updated: 3/15/2026, 6:17:33 PM
1class MedianFinder:
2    smaller = []
3    bigger = []
4
5    def __init__(self):
6        self.smaller = []
7        self.bigger = []
8
9    def addNum(self, num: int) -> None:
10        heapq.heappush(self.smaller, -num)
11
12    def findMedian(self) -> float:
13        while not (len(self.smaller) == len(self.bigger) + 1 or len(self.smaller) == len(self.bigger)):
14            top_smaller = self.smaller[0]
15            heapq.heappush(self.bigger, -top_smaller)
16            heapq.heappop(self.smaller)
17
18        while self.smaller and self.bigger and -self.smaller[0] > self.bigger[0]:
19            top_smaller = self.smaller[0]
20            top_bigger = self.bigger[0]
21            heapq.heappop(self.smaller)
22            heapq.heappop(self.bigger)
23            heapq.heappush(self.bigger, -top_smaller)
24            heapq.heappush(self.smaller, -top_bigger)
25        
26        # print("finding median")
27        # print(self.smaller)
28        # print(self.bigger)
29        if len(self.smaller) == len(self.bigger) + 1:
30            return -self.smaller[0]
31        
32        return (-self.smaller[0] + self.bigger[0]) / 2
33
34
35# Your MedianFinder object will be instantiated and called as such:
36# obj = MedianFinder()
37# obj.addNum(num)
38# param_2 = obj.findMedian()